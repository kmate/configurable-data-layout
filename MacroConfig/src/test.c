#include <boost/preprocessor.hpp>

// Prevent including malloc.h
#define ALLOCATE_FUNCTION malloc
#define DEALLOCATE_FUNCTION free
#include "macro_config.h"


// Test configurations
#define C_S SCALAR
#define C_PS PTR(SCALAR)
#define C_PPS PTR(PTR(SCALAR))
#define C_PPPS PTR(PTR(PTR(SCALAR)))
#define C_AS ARRAY(SCALAR)
#define C_PAS PTR(ARRAY(SCALAR))
#define C_PPAS PTR(PTR(ARRAY(SCALAR)))
#define C_APS ARRAY(PTR(SCALAR))
#define C_APPS ARRAY(PTR(PTR(SCALAR)))
#define C_PAPS PTR(ARRAY(PTR(SCALAR)))
#define C_PAPPS PTR(ARRAY(PTR(PTR(SCALAR))))
#define C_PPAPS PTR(PTR(ARRAY(PTR(SCALAR))))
#define C_PPAPPS PTR(PTR(ARRAY(PTR(PTR(SCALAR)))))
#define C_AAS ARRAY(ARRAY(SCALAR))
#define C_PAAS PTR(ARRAY(ARRAY(SCALAR)))
#define C_APAS ARRAY(PTR(ARRAY(SCALAR)))
#define C_AAPS ARRAY(ARRAY(PTR(SCALAR)))
#define C_APAPS ARRAY(PTR(ARRAY(PTR(SCALAR))))
#define C_PAPAS PTR(ARRAY(PTR(ARRAY(SCALAR))))
#define C_PAPAPS PTR(ARRAY(PTR(ARRAY(PTR(SCALAR)))))
#define C_PPAPPAPPS PTR(PTR(ARRAY(PTR(PTR(ARRAY(PTR(PTR(SCALAR))))))))

#define SCALAR_CONFIGURATIONS (C_S)(C_PS)(C_PPS)(C_PPPS)

#define ARRAY_CONFIGURATIONS  (C_AS)   (C_PAS)  (C_PPAS)  (C_APS)   (C_APPS)     (C_PAPS) \
                              (C_PAPPS)(C_PPAPS)(C_PPAPPS)(C_AAS)   (C_PAAS)     (C_APAS) \
                              (C_AAPS) (C_APAPS)(C_PAPAS) (C_PAPAPS)(C_PPAPPAPPS)

#define ARRAY_DIMENSIONS      ((2))    ((4))    ((8))     ((16))    ((32))       ((64))   \
                              ((2))    ((4))    ((8))     ((16,32)) ((2,4))      ((8,16)) \
                              ((32,2)) ((4,8))  ((16,32)) ((2,4))   ((8,16))

#define ALL_CONFIGURATIONS    SCALAR_CONFIGURATIONS ARRAY_CONFIGURATIONS


// Convert a test configuration into a label
#define MAKE_LABEL(prefix, config) BOOST_PP_CAT(prefix, BOOST_PP_CAT(__, BOOST_PP_TUPLE_ELEM(0, BOOST_PP_SEQ_FOLD_LEFT(MAKE_LABEL_, (BOOST_PP_EMPTY(), 0), config)))):
#define MAKE_LABEL_(s, state, item) \
    (BOOST_PP_CAT( \
        BOOST_PP_TUPLE_ELEM(0, state), \
        BOOST_PP_IIF( \
            BOOST_PP_GREATER(BOOST_PP_TUPLE_ELEM(1, state), 0), \
            BOOST_PP_CAT(_, item), \
            item \
        ) \
    ), BOOST_PP_INC(BOOST_PP_TUPLE_ELEM(1, state)))


// Configurable function arguments
#define ARGUMENT_WRAPPER(r, data, index, elem) \
/*  MAKE_LABEL(ARGUMENT, elem) */ \
    void BOOST_PP_CAT(f, index)(ARGUMENT( \
        BOOST_PP_TUPLE_ELEM(0, data), \
        BOOST_PP_TUPLE_ELEM(1, data), \
        elem \
    ));
BOOST_PP_SEQ_FOR_EACH_I(ARGUMENT_WRAPPER, (int, x), SCALAR_CONFIGURATIONS)
#undef ARGUMENT_WRAPPER

#define ARRAY_ARGUMENT_WRAPPER(r, data, index, elem) \
/*    MAKE_LABEL(ARRAY_ARGUMENT, elem) */ \
    void BOOST_PP_CAT(f, BOOST_PP_ADD(index, BOOST_PP_SEQ_SIZE(SCALAR_CONFIGURATIONS)))(ARRAY_ARGUMENT( \
        BOOST_PP_TUPLE_ELEM(0, data), \
        BOOST_PP_TUPLE_ELEM(1, data), \
        elem, \
        BOOST_PP_TUPLE_ENUM(BOOST_PP_SEQ_ELEM(index, ARRAY_DIMENSIONS)) \
    ));
BOOST_PP_SEQ_FOR_EACH_I(ARRAY_ARGUMENT_WRAPPER, (int, x), ARRAY_CONFIGURATIONS)
#undef ARRAY_ARGUMENT_WRAPPER


// Configurable function return value
#define DECLARE_FUNCTION_WRAPPER(r, data, index, elem) \
/*    MAKE_LABEL(DECLARE_FUNCTION, elem) */ \
    DECLARE_FUNCTION( \
        BOOST_PP_TUPLE_ELEM(0, data), \
        BOOST_PP_CAT(BOOST_PP_TUPLE_ELEM(1, data), index), \
        BOOST_PP_TUPLE_ELEM(2, data), \
        elem \
    );
BOOST_PP_SEQ_FOR_EACH_I(DECLARE_FUNCTION_WRAPPER, (int, g, (void)), ALL_CONFIGURATIONS)
#undef DECLARE_FUNCTION_WRAPPER


void test()
{
// Declaration
    #define DECLARE_WRAPPER(r, data, index, elem) \
        MAKE_LABEL(DECLARE, elem); \
        DECLARE( \
            BOOST_PP_TUPLE_ELEM(0, data), \
            BOOST_PP_CAT(BOOST_PP_TUPLE_ELEM(1, data), index), \
            elem \
        );
    BOOST_PP_SEQ_FOR_EACH_I(DECLARE_WRAPPER, (int, a), SCALAR_CONFIGURATIONS)
    #undef DECLARE_WRAPPER

    #define DECLARE_ARRAY_WRAPPER(r, data, index, elem) \
        MAKE_LABEL(DECLARE_ARRAY, elem); \
        DECLARE_ARRAY( \
            BOOST_PP_TUPLE_ELEM(0, data), \
            BOOST_PP_CAT(BOOST_PP_TUPLE_ELEM(1, data), BOOST_PP_ADD(index, BOOST_PP_SEQ_SIZE(SCALAR_CONFIGURATIONS))), \
            elem, \
            BOOST_PP_TUPLE_ENUM(BOOST_PP_SEQ_ELEM(index, ARRAY_DIMENSIONS)) \
        );
    BOOST_PP_SEQ_FOR_EACH_I(DECLARE_ARRAY_WRAPPER, (int, a), ARRAY_CONFIGURATIONS)
    #undef DECLARE_ARRAY_WRAPPER


// Allocation
    #define ALLOCATE_WRAPPER(r, data, index, elem) \
        MAKE_LABEL(ALLOCATE, elem) \
        ALLOCATE(BOOST_PP_CAT(data, index), elem);
    BOOST_PP_SEQ_FOR_EACH_I(ALLOCATE_WRAPPER, a, ALL_CONFIGURATIONS)
    #undef ALLOCATE_WRAPPER


// Value of
    #define VALUE_OF_WRAPPER(r, data, index, elem) \
        MAKE_LABEL(VALUE_OF, elem) \
        VALUE_OF(BOOST_PP_CAT(data, index), elem);
    BOOST_PP_SEQ_FOR_EACH_I(VALUE_OF_WRAPPER, a, ALL_CONFIGURATIONS)
    #undef VALUE_OF_WRAPPER


// Value at a specified array index
    #define VALUE_AT_WRAPPER(r, data, index, elem) \
        MAKE_LABEL(VALUE_AT, elem) \
        VALUE_AT( \
            BOOST_PP_CAT(data, BOOST_PP_ADD(index, BOOST_PP_SEQ_SIZE(SCALAR_CONFIGURATIONS))), \
            elem, \
            BOOST_PP_TUPLE_ENUM(BOOST_PP_SEQ_ELEM(index, ARRAY_DIMENSIONS)) \
        );
    BOOST_PP_SEQ_FOR_EACH_I(VALUE_AT_WRAPPER, a, ARRAY_CONFIGURATIONS)
    #undef VALUE_AT_WRAPPER

// Deallocation
    #define DEALLOCATE_WRAPPER(r, data, index, elem) \
        MAKE_LABEL(DEALLOCATE, elem) \
        DEALLOCATE(BOOST_PP_CAT(data, index), elem);
    BOOST_PP_SEQ_FOR_EACH_I(DEALLOCATE_WRAPPER, a, ALL_CONFIGURATIONS)
    #undef DEALLOCATE_WRAPPER
}
