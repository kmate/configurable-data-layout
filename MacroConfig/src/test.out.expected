void f0(int x); void f1(int * x); void f2(int ** x); void f3(int *** x);
void f4(int ( x[2])); void f5(int ((* x) [4])); void f6(int ((*(* x) ) [8])); void f7(int (*( x[16]) )); void f8(int (*(*( x[32]) ) )); void f9(int (*((* x) [64]) )); void f10(int (*(*((* x) [2]) ) )); void f11(int (*((*(* x) ) [4]) )); void f12(int (*(*((*(* x) ) [8]) ) )); void f13(int (( x[16]) [32])); void f14(int (((* x) [2]) [4])); void f15(int ((*( x[8]) ) [16])); void f16(int (*(( x[32]) [2]) )); void f17(int (*((*( x[4]) ) [8]) )); void f18(int ((*((* x) [16]) ) [32])); void f19(int (*((*((* x) [2]) ) [4]) )); void f20(int (*(*((*(*((*(* x) ) [8]) ) ) [16]) ) ));
int g0 (void); int * g1 (void); int ** g2 (void); int *** g3 (void); int * g4 (void); int ** g5 (void); int *** g6 (void); int ** g7 (void); int *** g8 (void); int *** g9 (void); int **** g10 (void); int **** g11 (void); int ***** g12 (void); int ** g13 (void); int *** g14 (void); int *** g15 (void); int *** g16 (void); int **** g17 (void); int **** g18 (void); int ***** g19 (void); int ******** g20 (void);
void test()
{
    DECLARE_IMPL__SCALAR:; int a0; DECLARE_IMPL__PTR_SCALAR:; int * a1; DECLARE_IMPL__PTR_PTR_SCALAR:; int ** a2; DECLARE_IMPL__PTR_PTR_PTR_SCALAR:; int *** a3;
    DECLARE_ARRAY_IMPL__ARRAY_SCALAR:; int ( a4[2]); DECLARE_ARRAY_IMPL__PTR_ARRAY_SCALAR:; int ((* a5) [4]); DECLARE_ARRAY_IMPL__PTR_PTR_ARRAY_SCALAR:; int ((*(* a6) ) [8]); DECLARE_ARRAY_IMPL__ARRAY_PTR_SCALAR:; int (*( a7[16]) ); DECLARE_ARRAY_IMPL__ARRAY_PTR_PTR_SCALAR:; int (*(*( a8[32]) ) ); DECLARE_ARRAY_IMPL__PTR_ARRAY_PTR_SCALAR:; int (*((* a9) [64]) ); DECLARE_ARRAY_IMPL__PTR_ARRAY_PTR_PTR_SCALAR:; int (*(*((* a10) [2]) ) ); DECLARE_ARRAY_IMPL__PTR_PTR_ARRAY_PTR_SCALAR:; int (*((*(* a11) ) [4]) ); DECLARE_ARRAY_IMPL__PTR_PTR_ARRAY_PTR_PTR_SCALAR:; int (*(*((*(* a12) ) [8]) ) ); DECLARE_ARRAY_IMPL__ARRAY_ARRAY_SCALAR:; int (( a13[16]) [32]); DECLARE_ARRAY_IMPL__PTR_ARRAY_ARRAY_SCALAR:; int (((* a14) [2]) [4]); DECLARE_ARRAY_IMPL__ARRAY_PTR_ARRAY_SCALAR:; int ((*( a15[8]) ) [16]); DECLARE_ARRAY_IMPL__ARRAY_ARRAY_PTR_SCALAR:; int (*(( a16[32]) [2]) ); DECLARE_ARRAY_IMPL__ARRAY_PTR_ARRAY_PTR_SCALAR:; int (*((*( a17[4]) ) [8]) ); DECLARE_ARRAY_IMPL__PTR_ARRAY_PTR_ARRAY_SCALAR:; int ((*((* a18) [16]) ) [32]); DECLARE_ARRAY_IMPL__PTR_ARRAY_PTR_ARRAY_PTR_SCALAR:; int (*((*((* a19) [2]) ) [4]) ); DECLARE_ARRAY_IMPL__PTR_PTR_ARRAY_PTR_PTR_ARRAY_PTR_PTR_SCALAR:; int (*(*((*(*((*(* a20) ) [8]) ) ) [16]) ) );
    ALLOCATE_IMPL__SCALAR: do { } while(0); ALLOCATE_IMPL__PTR_SCALAR: do { a1 = malloc(sizeof(*(a1))); } while(0); ALLOCATE_IMPL__PTR_PTR_SCALAR: do { a2 = malloc(sizeof(*(a2))); *(a2) = malloc(sizeof(*(*(a2)))); } while(0); ALLOCATE_IMPL__PTR_PTR_PTR_SCALAR: do { a3 = malloc(sizeof(*(a3))); *(a3) = malloc(sizeof(*(*(a3)))); *(*(a3)) = malloc(sizeof(*(*(*(a3))))); } while(0); ALLOCATE_IMPL__ARRAY_SCALAR: do { } while(0); ALLOCATE_IMPL__PTR_ARRAY_SCALAR: do { a5 = malloc(sizeof(*(a5))); } while(0); ALLOCATE_IMPL__PTR_PTR_ARRAY_SCALAR: do { a6 = malloc(sizeof(*(a6))); *(a6) = malloc(sizeof(*(*(a6)))); } while(0); ALLOCATE_IMPL__ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a7) / sizeof((a7)[0]); ++i0) { (a7)[i0] = malloc(sizeof(*((a7)[i0]))); } } while(0); ALLOCATE_IMPL__ARRAY_PTR_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a8) / sizeof((a8)[0]); ++i0) { (a8)[i0] = malloc(sizeof(*((a8)[i0]))); *((a8)[i0]) = malloc(sizeof(*(*((a8)[i0])))); } } while(0); ALLOCATE_IMPL__PTR_ARRAY_PTR_SCALAR: do { a9 = malloc(sizeof(*(a9))); for(int i0 = 0; i0 < sizeof(*(a9)) / sizeof((*(a9))[0]); ++i0) { (*(a9))[i0] = malloc(sizeof(*((*(a9))[i0]))); } } while(0); ALLOCATE_IMPL__PTR_ARRAY_PTR_PTR_SCALAR: do { a10 = malloc(sizeof(*(a10))); for(int i0 = 0; i0 < sizeof(*(a10)) / sizeof((*(a10))[0]); ++i0) { (*(a10))[i0] = malloc(sizeof(*((*(a10))[i0]))); *((*(a10))[i0]) = malloc(sizeof(*(*((*(a10))[i0])))); } } while(0); ALLOCATE_IMPL__PTR_PTR_ARRAY_PTR_SCALAR: do { a11 = malloc(sizeof(*(a11))); *(a11) = malloc(sizeof(*(*(a11)))); for(int i0 = 0; i0 < sizeof(*(*(a11))) / sizeof((*(*(a11)))[0]); ++i0) { (*(*(a11)))[i0] = malloc(sizeof(*((*(*(a11)))[i0]))); } } while(0); ALLOCATE_IMPL__PTR_PTR_ARRAY_PTR_PTR_SCALAR: do { a12 = malloc(sizeof(*(a12))); *(a12) = malloc(sizeof(*(*(a12)))); for(int i0 = 0; i0 < sizeof(*(*(a12))) / sizeof((*(*(a12)))[0]); ++i0) { (*(*(a12)))[i0] = malloc(sizeof(*((*(*(a12)))[i0]))); *((*(*(a12)))[i0]) = malloc(sizeof(*(*((*(*(a12)))[i0])))); } } while(0); ALLOCATE_IMPL__ARRAY_ARRAY_SCALAR: do { } while(0); ALLOCATE_IMPL__PTR_ARRAY_ARRAY_SCALAR: do { a14 = malloc(sizeof(*(a14))); } while(0); ALLOCATE_IMPL__ARRAY_PTR_ARRAY_SCALAR: do { for(int i0 = 0; i0 < sizeof(a15) / sizeof((a15)[0]); ++i0) { (a15)[i0] = malloc(sizeof(*((a15)[i0]))); } } while(0); ALLOCATE_IMPL__ARRAY_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a16) / sizeof((a16)[0]); ++i0) { for(int i1 = 0; i1 < sizeof((a16)[i0]) / sizeof(((a16)[i0])[0]); ++i1) { ((a16)[i0])[i1] = malloc(sizeof(*(((a16)[i0])[i1]))); } } } while(0); ALLOCATE_IMPL__ARRAY_PTR_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a17) / sizeof((a17)[0]); ++i0) { (a17)[i0] = malloc(sizeof(*((a17)[i0]))); for(int i1 = 0; i1 < sizeof(*((a17)[i0])) / sizeof((*((a17)[i0]))[0]); ++i1) { (*((a17)[i0]))[i1] = malloc(sizeof(*((*((a17)[i0]))[i1]))); } } } while(0); ALLOCATE_IMPL__PTR_ARRAY_PTR_ARRAY_SCALAR: do { a18 = malloc(sizeof(*(a18))); for(int i0 = 0; i0 < sizeof(*(a18)) / sizeof((*(a18))[0]); ++i0) { (*(a18))[i0] = malloc(sizeof(*((*(a18))[i0]))); } } while(0); ALLOCATE_IMPL__PTR_ARRAY_PTR_ARRAY_PTR_SCALAR: do { a19 = malloc(sizeof(*(a19))); for(int i0 = 0; i0 < sizeof(*(a19)) / sizeof((*(a19))[0]); ++i0) { (*(a19))[i0] = malloc(sizeof(*((*(a19))[i0]))); for(int i1 = 0; i1 < sizeof(*((*(a19))[i0])) / sizeof((*((*(a19))[i0]))[0]); ++i1) { (*((*(a19))[i0]))[i1] = malloc(sizeof(*((*((*(a19))[i0]))[i1]))); } } } while(0); ALLOCATE_IMPL__PTR_PTR_ARRAY_PTR_PTR_ARRAY_PTR_PTR_SCALAR: do { a20 = malloc(sizeof(*(a20))); *(a20) = malloc(sizeof(*(*(a20)))); for(int i0 = 0; i0 < sizeof(*(*(a20))) / sizeof((*(*(a20)))[0]); ++i0) { (*(*(a20)))[i0] = malloc(sizeof(*((*(*(a20)))[i0]))); *((*(*(a20)))[i0]) = malloc(sizeof(*(*((*(*(a20)))[i0])))); for(int i1 = 0; i1 < sizeof(*(*((*(*(a20)))[i0]))) / sizeof((*(*((*(*(a20)))[i0])))[0]); ++i1) { (*(*((*(*(a20)))[i0])))[i1] = malloc(sizeof(*((*(*((*(*(a20)))[i0])))[i1]))); *((*(*((*(*(a20)))[i0])))[i1]) = malloc(sizeof(*(*((*(*((*(*(a20)))[i0])))[i1])))); } } } while(0);
    VALUE_OF_IMPL__SCALAR: (a0); VALUE_OF_IMPL__PTR_SCALAR: *(a1); VALUE_OF_IMPL__PTR_PTR_SCALAR: **(a2); VALUE_OF_IMPL__PTR_PTR_PTR_SCALAR: ***(a3); VALUE_OF_IMPL__ARRAY_SCALAR: (a4); VALUE_OF_IMPL__PTR_ARRAY_SCALAR: *(a5); VALUE_OF_IMPL__PTR_PTR_ARRAY_SCALAR: **(a6); VALUE_OF_IMPL__ARRAY_PTR_SCALAR: (a7); VALUE_OF_IMPL__ARRAY_PTR_PTR_SCALAR: (a8); VALUE_OF_IMPL__PTR_ARRAY_PTR_SCALAR: *(a9); VALUE_OF_IMPL__PTR_ARRAY_PTR_PTR_SCALAR: *(a10); VALUE_OF_IMPL__PTR_PTR_ARRAY_PTR_SCALAR: **(a11); VALUE_OF_IMPL__PTR_PTR_ARRAY_PTR_PTR_SCALAR: **(a12); VALUE_OF_IMPL__ARRAY_ARRAY_SCALAR: (a13); VALUE_OF_IMPL__PTR_ARRAY_ARRAY_SCALAR: *(a14); VALUE_OF_IMPL__ARRAY_PTR_ARRAY_SCALAR: (a15); VALUE_OF_IMPL__ARRAY_ARRAY_PTR_SCALAR: (a16); VALUE_OF_IMPL__ARRAY_PTR_ARRAY_PTR_SCALAR: (a17); VALUE_OF_IMPL__PTR_ARRAY_PTR_ARRAY_SCALAR: *(a18); VALUE_OF_IMPL__PTR_ARRAY_PTR_ARRAY_PTR_SCALAR: *(a19); VALUE_OF_IMPL__PTR_PTR_ARRAY_PTR_PTR_ARRAY_PTR_PTR_SCALAR: **(a20);
    VALUE_AT_IMPL__ARRAY_SCALAR: ((a4))[2]; VALUE_AT_IMPL__PTR_ARRAY_SCALAR: (*(a5))[4]; VALUE_AT_IMPL__PTR_PTR_ARRAY_SCALAR: (**(a6))[8]; VALUE_AT_IMPL__ARRAY_PTR_SCALAR: *((a7))[16]; VALUE_AT_IMPL__ARRAY_PTR_PTR_SCALAR: **((a8))[32]; VALUE_AT_IMPL__PTR_ARRAY_PTR_SCALAR: *(*(a9))[64]; VALUE_AT_IMPL__PTR_ARRAY_PTR_PTR_SCALAR: **(*(a10))[2]; VALUE_AT_IMPL__PTR_PTR_ARRAY_PTR_SCALAR: *(**(a11))[4]; VALUE_AT_IMPL__PTR_PTR_ARRAY_PTR_PTR_SCALAR: **(**(a12))[8]; VALUE_AT_IMPL__ARRAY_ARRAY_SCALAR: (((a13))[16])[32]; VALUE_AT_IMPL__PTR_ARRAY_ARRAY_SCALAR: ((*(a14))[2])[4]; VALUE_AT_IMPL__ARRAY_PTR_ARRAY_SCALAR: (*((a15))[8])[16]; VALUE_AT_IMPL__ARRAY_ARRAY_PTR_SCALAR: *(((a16))[32])[2]; VALUE_AT_IMPL__ARRAY_PTR_ARRAY_PTR_SCALAR: *(*((a17))[4])[8]; VALUE_AT_IMPL__PTR_ARRAY_PTR_ARRAY_SCALAR: (*(*(a18))[16])[32]; VALUE_AT_IMPL__PTR_ARRAY_PTR_ARRAY_PTR_SCALAR: *(*(*(a19))[2])[4]; VALUE_AT_IMPL__PTR_PTR_ARRAY_PTR_PTR_ARRAY_PTR_PTR_SCALAR: **(**(**(a20))[8])[16];
    DEALLOCATE_IMPL__SCALAR: do { } while(0); DEALLOCATE_IMPL__PTR_SCALAR: do { free(a1); } while(0); DEALLOCATE_IMPL__PTR_PTR_SCALAR: do { free(*(a2)); free(a2); } while(0); DEALLOCATE_IMPL__PTR_PTR_PTR_SCALAR: do { free(*(*(a3))); free(*(a3)); free(a3); } while(0); DEALLOCATE_IMPL__ARRAY_SCALAR: do { } while(0); DEALLOCATE_IMPL__PTR_ARRAY_SCALAR: do { free(a5); } while(0); DEALLOCATE_IMPL__PTR_PTR_ARRAY_SCALAR: do { free(*(a6)); free(a6); } while(0); DEALLOCATE_IMPL__ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a7) / sizeof((a7)[0]); ++i0) { free((a7)[i0]); } } while(0); DEALLOCATE_IMPL__ARRAY_PTR_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a8) / sizeof((a8)[0]); ++i0) { free(*((a8)[i0])); free((a8)[i0]); } } while(0); DEALLOCATE_IMPL__PTR_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(a9)) / sizeof((*(a9))[0]); ++i0) { free((*(a9))[i0]); } free(a9); } while(0); DEALLOCATE_IMPL__PTR_ARRAY_PTR_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(a10)) / sizeof((*(a10))[0]); ++i0) { free(*((*(a10))[i0])); free((*(a10))[i0]); } free(a10); } while(0); DEALLOCATE_IMPL__PTR_PTR_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(*(a11))) / sizeof((*(*(a11)))[0]); ++i0) { free((*(*(a11)))[i0]); } free(*(a11)); free(a11); } while(0); DEALLOCATE_IMPL__PTR_PTR_ARRAY_PTR_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(*(a12))) / sizeof((*(*(a12)))[0]); ++i0) { free(*((*(*(a12)))[i0])); free((*(*(a12)))[i0]); } free(*(a12)); free(a12); } while(0); DEALLOCATE_IMPL__ARRAY_ARRAY_SCALAR: do { } while(0); DEALLOCATE_IMPL__PTR_ARRAY_ARRAY_SCALAR: do { free(a14); } while(0); DEALLOCATE_IMPL__ARRAY_PTR_ARRAY_SCALAR: do { for(int i0 = 0; i0 < sizeof(a15) / sizeof((a15)[0]); ++i0) { free((a15)[i0]); } } while(0); DEALLOCATE_IMPL__ARRAY_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a16) / sizeof((a16)[0]); ++i0) { for(int i1 = 0; i1 < sizeof((a16)[i0]) / sizeof(((a16)[i0])[0]); ++i1) { free(((a16)[i0])[i1]); } } } while(0); DEALLOCATE_IMPL__ARRAY_PTR_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(a17) / sizeof((a17)[0]); ++i0) { for(int i1 = 0; i1 < sizeof(*((a17)[i0])) / sizeof((*((a17)[i0]))[0]); ++i1) { free((*((a17)[i0]))[i1]); } free((a17)[i0]); } } while(0); DEALLOCATE_IMPL__PTR_ARRAY_PTR_ARRAY_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(a18)) / sizeof((*(a18))[0]); ++i0) { free((*(a18))[i0]); } free(a18); } while(0); DEALLOCATE_IMPL__PTR_ARRAY_PTR_ARRAY_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(a19)) / sizeof((*(a19))[0]); ++i0) { for(int i1 = 0; i1 < sizeof(*((*(a19))[i0])) / sizeof((*((*(a19))[i0]))[0]); ++i1) { free((*((*(a19))[i0]))[i1]); } free((*(a19))[i0]); } free(a19); } while(0); DEALLOCATE_IMPL__PTR_PTR_ARRAY_PTR_PTR_ARRAY_PTR_PTR_SCALAR: do { for(int i0 = 0; i0 < sizeof(*(*(a20))) / sizeof((*(*(a20)))[0]); ++i0) { for(int i1 = 0; i1 < sizeof(*(*((*(*(a20)))[i0]))) / sizeof((*(*((*(*(a20)))[i0])))[0]); ++i1) { free(*((*(*((*(*(a20)))[i0])))[i1])); free((*(*((*(*(a20)))[i0])))[i1]); } free(*((*(*(a20)))[i0])); free((*(*(a20)))[i0]); } free(*(a20)); free(a20); } while(0);
}
